[HotFix] [SVCS-889] Keep "REMOTE_USER" check in place for auth ID purpose

cslzchen opened the pull request:
## Ticket

https://openscience.atlassian.net/browse/SVCS-889

## Purpose

### Main

Keep "REMOTE_USER" check in place for auth ID purpose

The header "REMOTE_USER" is still REQUIRED as an id for institution users although 1) OSF and CAS already rely on `username` for ID purpose and 2) `notifyRemotePrincipalAuthenticated()` guarantees that `username` is provided. The reason for keeping this extra check in place is that `username` is not always the identifier since it can be `eppn`, `mail` or other attributes. "REMOTE_USER" is defined as `REMOTE_USER="eppn uid persistent-id targeted-id"`, in which very
attribute can be considered as the users' institution identity. In short, CAS requires at least one of the four ID attributes `eppn`, `uid`, `persistent-id`, `targeted-id` in addition to attributes that are mapped to `username` and `fullname`. Please see our Shibboleth server's configuration (private) for detailed mapping for attributes. The following link provides the best practice of using "REMOTE_USER": https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPAttributeAccess

### More on **RMOTE_USER**

Here is a few quotes from the link provided above.

> "REMOTE_USER" is a special environment variable established by the CGI for use in passing the "authenticated username" to applications. It is NOT a header, and is designed to be set only when the web server is informed by a module that the user accessing the application is known. It is empty or unset otherwise.

> In some cases, the user's IdP may not provide enough information to uniquely identify the user, but instead supply only an attribute intended to grant the user access to something. Or it's possible that the user's own privacy requirements preclude supplying identifying information. In either case, the application might still have to run but will not have a username to key off of. This can certainly be treated as an error by the application. What it can't do is crash the application (unless you're fine with people thinking your site is designed by incompetents).

As discussed with @icereval , OSF institution login requires the user's institution identity. We choose to **treat the case as error** when attributes for "REMOTE_USER" is not provided.


## Changes

Only comments update.

## Side effects

No

## QA Notes

No

## Deployment Notes

No

cslzchen committed:
Remove requirements for the header "REMOTE_USER" for SAML/SHIB login

- The header "REMOTE_USER" is not required as an identifier for
  institution users. OSF and CAS rely on `username` to identify
  such  users. The method `notifyRemotePrincipalAuthenticated()`
  guarantees that `username` is provided. For most cases, it is
  either the `eppn` or `mail` attribute.
- In addition, update the comments for SAML/SHIB-based login

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -273,34 +273,46 @@ public void afterPropertiesSet() throws Exception {
+                logger.info("Header \"REMOTE_USR\" is empty.");
+                logger.debug("Header \"REMOTE_USR\" is only available if the institution has provided"
+                        + "any of the three attributes: ``eppn``, ``persistent-id`` or ``targeted-id``");
+                remoteUser = "";
```
If the header "REMOTE_USER" does not exit, log the incident and set `remoteUser` to the empty string.

cslzchen reviewed (commented):
cc @icereval 

PR ready 🎆 . I will hot-fix this in soon.

icereval reviewed (dismissed):
Can you add to the PR a link and quote for the article referencing the justification for removing the Remote User check.

icereval reviewed (changes requested):
After further discussion, lets keep this as a just in case check.

cslzchen committed:
Reinstate "REMOTE_USER" as a auth ID requirement

The header "REMOTE_USER" is still REQUIRED as an id for institution
users although 1) OSF and CAS already rely on `username` for ID pur-
pose and 2) `notifyRemotePrincipalAuthenticated()` guarantees that
`username` is provided. The reason for keeping this extra check in
place is that `username` is not always the identifier since it can
be `eppn`, `mail` or other attributes. "REMOTE_USER" is defined as
`REMOTE_USER="eppn uid persistent-id targeted-id"`, in which every
attribute can be considered as the users' institution identity. In
short, CAS requires at least one of the four ID attributes `eppn`,
`uid`, `persistent-id`, `targeted-id` in addition to attributes that
are mapped to `username` and `fullname`. Please see our Shibboleth
server's configuration (private) for detailed mapping for attributes.
The following link provides the best practice of using "REMOTE_USER":
wiki.shibboleth.net/confluence/display/SHIB2/NativeSPAttributeAccess

cslzchen committed:
Add a TODO comment for delaying the "REMOTE_USER" exception

When the exception is thrown, neither the logs nor sentry knows which
the institution is or who the user is. If we can delay it until we
have the institution id and username, the logs will be more helpful.

cslzchen commented:
Update: comments update only, no longer functional, rebase to target develop. Closing.
